mksusecd

```diff
@@ -3122,105 +3128,171 @@ sub get_archive_type
+      # But only if part number matches or is 0 (== all parts).
+      if(!$p && ($part == 0 || $part == $cnt)) {
+        open $p, "| ( cd $dst ; $cpio_cmd )" or die "failed to open cpio: $!\n";
+        $write_ofs = 0;
+      }
 
-    return $read;
-  };
+      # Write data and track output size for padding calculation at the end.
+      if($p) {
+        syswrite $p, $buf;
+        $write_ofs += length $buf;
+      }
+    }
 
-  if(open $f, $file) {
-    while(($len = &$read_write(110)) == 110) {
-      $ofs += 110;
+    return length $buf;
+  }
+
+  # Write padding bytes (pad with 0 to full 512 byte blocks) and close
+  # output pipe.
+  #
+  # write_pad_and_close()
+  #
+  # This also sets a sync flag indicating that we should search for the next
+  # valid cpio header in the input stream.
+  #
+  sub write_pad_and_close
+  {
+    if($p) {
+      my $pad = (($write_ofs + 0x1ff) & ~0x1ff) - $write_ofs;
+      syswrite $p, "\x00" x $pad, $pad if $pad;
+      close $p;
+      undef $p;
+    }
 
-      # printf "header = \"%s\" %d\n", $buf, length($buf);
+    # search for next cpio header in input stream
+    $sync = 1;
+  }
 
-      $magic = substr($buf, 0, 6);
-      $head = substr($buf, 6);
+  # open archive and get going...
+  if(open $f, $file) {
+    my $len;
 
-      $fname_len = hex substr $buf, 94, 8;
-      $data_len = hex substr $buf, 54, 8;
+    # We have to trace the cpio archive structure.
+    # Keep going as long as there's a header.
+    while(($len = read_write(110)) == 110) {
+      my $magic = substr($buf, 0, 6);
+      my $head = substr($buf, 6);
 
-      # print "$magic - $fname_len - $data_len\n";
+      my $fname_len = hex substr $buf, 94, 8;
+      my $data_len = hex substr $buf, 54, 8;
 
-      die "broken cpio header\n" if $magic ne "070701" && $magic ne "070702";
+      die "broken cpio header\n" if $magic !~ /^07070[12]$/;
 
       $fname_len += (2, 1, 0, 3)[$fname_len & 3];
       $data_len = (($data_len + 3) & ~3);
 
-      &$read_write($fname_len);
-      $fname = $buf;
-      $fname =~ s/\x00*$//;
-
-      &$read_write($data_len);
-
-      $ofs += $fname_len + $data_len;
+      read_write $fname_len;
 
-      # print "fname = \"$fname\"\n";
+      my $fname = $buf;
+      $fname =~ s/\x00*$//;
 
-      # printf "ofs = 0x%x\n", $ofs;
+      read_write $data_len;
 
+      # Look for cpio archive end marker.
+      # If found, close cpio process. A new process will be started at the
+      # next valid cpio archive header.
       if(
         $fname eq 'TRAILER!!!' &&
-        $head eq '00000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000B00000000'
+        $head =~ /^0{39}10{55}b0{8}$/i
       ) {
-        my $pad = ($ofs + 0x1ff) & ~0x1ff;
-        &$read_write($pad - $ofs);
-        $ofs = $pad;
-        $cnt++;
-        close $p if $p;
-        undef $p;
+        write_pad_and_close;
+        # exit if we're done
+        close $f, return if $cnt++ == $part;
```

Comment:
please use

    if ($cnt++ == $part)
    {
        close $f;
        return;
    }

That's much more natural to read IMHO.